packages/storage/compat/service.ts

```diff
@@ -0,0 +1,121 @@
+/**
+ * @license
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as types from '@firebase/storage-types';
+import { StorageService, isUrl, ref } from '../src/service';
+import { Location } from '../src/implementation/location';
+import * as args from '../src/implementation/args';
+import { ReferenceCompat } from './reference';
+import { Reference } from '../src/reference';
+
+export function urlValidator(maybeUrl: unknown): void {
+  if (typeof maybeUrl !== 'string') {
+    throw 'Path is not a string.';
+  }
+  if (!isUrl) {
+    throw 'Expected full URL but got a child path, use ref instead.';
+  }
+  try {
+    Location.makeFromUrl(maybeUrl as string);
+  } catch (e) {
+    throw 'Expected valid full URL but got an invalid one.';
+  }
+}
+
+export function pathValidator(path: unknown): void {
+  if (typeof path !== 'string') {
+    throw 'Path is not a string.';
+  }
+  if (isUrl(path)) {
+    throw 'Expected child path but got a URL, use refFromURL instead.';
+  }
+}
+
+/**
+ * A service that provides firebaseStorage.Reference instances.
+ * @param opt_url gs:// url to a custom Storage Bucket
+ */
+export class StorageServiceCompat implements types.FirebaseStorage {
+  constructor(
+    readonly delegate: StorageService,
+    private converter: (ref: Reference) => ReferenceCompat
+  ) {}
+
+  app = this.delegate.app;
+  maxOperationRetryTime = this.delegate.maxOperationRetryTime;
+  maxUploadRetryTime = this.delegate.maxUploadRetryTime;
+  INTERNAL = new ServiceInternals(this);
```

Comment:
Compat services are implementing this [interface](https://github.com/firebase/firebase-js-sdk/blob/master/packages/app-types/private.d.ts#L38) for now, which doesn't define a `_delete()` method. We can add it, but the current interface seems to work fine. I would change to just use a plain object instead of a class though.

Reply:
Changed to a plain object.